Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm

```diff
@@ -572,31 +587,41 @@
 
     bool paintedATile = false;
 
-    auto tilingScaleFactor = 1.0f;
-    if (auto* tiledBacking = m_pdfContentsLayer->tiledBacking())
-        tilingScaleFactor = tiledBacking->tilingScaleFactor();
-
-    // This scale takes us from "painting" coordinates into the coordinate system of the tile grid,
-    // so we can paint tiles directly.
-    auto scaleTransform = tileToPaintingTransform(tilingScaleFactor);
-    {
+    auto paintTilesIntersectingWithPageBounds = [&pageBoundsInPaintingCoordinates, &paintedATile](GraphicsContext& context, HashMap<TileForGrid, RenderedTile>& tiles, float tilingScaleFactor) {
+        auto scaleTransform = tileToPaintingTransform(tilingScaleFactor);
         auto stateSaver = GraphicsContextStateSaver(context);
         context.concatCTM(scaleTransform);
 
-        for (auto& keyValuePair : m_rendereredTiles) {
-            auto& renderedTile = keyValuePair.value;
-
+        for (auto& [tileInfo, renderedTile] : tiles) {
             auto tileClipInPaintingCoordinates = scaleTransform.mapRect(renderedTile.tileInfo.tileRect);
             if (!pageBoundsInPaintingCoordinates.intersects(tileClipInPaintingCoordinates))
                 continue;
 
-            LOG_WITH_STREAM(PDFAsyncRendering, stream << "AsyncPDFRenderer::paintTilesForPage " << pageBoundsInPaintingCoordinates  << " - painting tile for " << keyValuePair.key << " with clip " << renderedTile.tileInfo.tileRect << " tiling scale " << tilingScaleFactor);
-
+            LOG_WITH_STREAM(PDFAsyncRendering, stream << "AsyncPDFRenderer::paintTilesForPage " << pageBoundsInPaintingCoordinates  << " - painting tile for " << tileInfo << " with clip " << renderedTile.tileInfo.tileRect << " tiling scale " << tilingScaleFactor);
             context.drawImageBuffer(*renderedTile.buffer, renderedTile.tileInfo.tileRect.location());
             paintedATile = true;
         }
+    };
+
+    auto* tiledBacking = m_pdfContentsLayer->tiledBacking();
+    if (m_staleTileState && tiledBacking) {
+        Vector<TileIndex> renderedTileIndices;
+        for (auto& [tileInfo, renderedTile] : m_rendereredTiles)
+            renderedTileIndices.append(tileInfo.tileIndex);
+
+        if (tiledBacking->containsCoverageRect(renderedTileIndices))
+            m_staleTileState.reset();
     }
 
+    if (!m_staleTileState) {
+        auto tilingScaleFactor = 1.0f;
+        if (tiledBacking)
+            tilingScaleFactor = tiledBacking->tilingScaleFactor();
+
+        paintTilesIntersectingWithPageBounds(context, m_rendereredTiles, tilingScaleFactor);
+    } else
+        paintTilesIntersectingWithPageBounds(context, m_staleTileState->staleTiles(), m_staleTileState->tilingScaleFactor());
```

Comment:
It looks like you need to call `paintTilesIntersectingWithPageBounds` regardless, so maybe refactor this a little bit and write one call instead.